sui-execution/latest/sui-move-natives/src/eth_state_proof.rs

```diff
@@ -93,17 +114,64 @@ pub fn verify_message_proof(
     let message_map_index = get_message_storage_slot(message.clone(), dwallet_id.clone(), map_slot)
         .map_err(|_| PartialVMError::new(StatusCode::UNKNOWN_STATUS))?;
 
-    let contract_address = Address::from_slice(contract_address.as_slice());
-    let state_root = Bytes32::try_from(state_root.as_slice())
+    let contract_address = String::from_utf8(contract_address.clone())
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let contract_address: Address = contract_address
+        .parse()
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let mut beacon_block: BeaconBlock = serde_json::from_slice(&beacon_block)
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let beacon_block_type = String::from_utf8(beacon_block_type.clone())
         .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
 
+    let beacon_block_type: BeaconBlockType = BeaconBlockType::from_str(&beacon_block_type)
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let mut beacon_block_body =
+        BeaconBlockBodyWrapper::new_from_json(beacon_block_body, &beacon_block_type)
+            .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+    let beacon_block_body: BeaconBlockBody = beacon_block_body.inner();
+
+    let beacon_block_execution_payload =
+        ExecutionPayloadWrapper::new_from_json(beacon_block_execution_payload, &beacon_block_type)
+            .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+    let beacon_block_execution_payload: ExecutionPayload = beacon_block_execution_payload.inner();
+
+    let beacon_block_body = BeaconBlockBody::new_from_existing_with_execution_payload(
+        beacon_block_body,
+        beacon_block_execution_payload,
+    );
+    beacon_block.body = beacon_block_body;
+
+    let eth_state_data = bcs::from_bytes::<ConsensusStateManager<NimbusRpc>>(&eth_state_data)
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let beacon_block_hash = beacon_block
+        .hash_tree_root()
+        .map_err(|_| PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))?;
+
+    let mut verified_block_header = eth_state_data.get_finalized_header();
+
+    let verified_block_hash = verified_block_header
+        .hash_tree_root()
+        .map_err(|_| PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))?;
+
+    if beacon_block_hash != verified_block_hash {
```

Comment:
don't you need to actually compute the hash yourself?

Reply:
This is what I do.
method call `.hash_tree_root()` gives us the hash of the block/header.

I compare the hash of the input beacon block, and the hash of the verified block header that is saved in the `EthereumState` object.

Reply:
I'm not sure that this method computes the hash, it might retrtieve it from some field. Can you link the implementation?

Reply:
https://docs.rs/ssz_rs/latest/ssz_rs/trait.Merkleized.html

`Merkleized` is a trait that is being implemented in our code by using the `SimpleSerialize` derive, provided by `ssz_rs` lib.

Reply:
You can also see it's specs here:
[ssz specs](https://github.com/ethereum/consensus-specs/blob/fa09d896484bbe240334fa21ffaa454bafe5842e/ssz/simple-serialize.md#merkleization)

[ssz readme](https://github.com/ralexstokes/ssz-rs/blob/1df4cd9b849a48c44a6105abc3a38d21cd4fd8d3/README.md?plain=1#L29)

Reply:
So definitely the Merkle roots get computed and it makes sense that the header contains just the hash of the beacon body. Two things are left for us to check:
1. can we access the beacon body hash from the verified header directly, and then hash the block ourselves to see if it matches (this probably is an overkill, but I'm good with doing excess checks)?
I think we can, using `finalized_header.body_root` and comparing it to the hash of the `beacon_block_body`. let's try to add this check
2. I'm not sure about the data structure of the verified header - it should be a `BeaconBlockHeader`, but I see it is a `Header`: `finalized_header: Header`. I see its not only a naming issue, but these two types exist. They do seem to hold the exact same values, so there could be some overloading here, but I'm not sure that the `finalized_header` is necessairly a `BeaconBlockHeader` yet

Reply:
In fact, `BeaconBlockHeader` is not actually used anywhere in the code, only `Header` does. 
So that leaves the question: do you know (and can point me out) to where in the protocol it says the header of the `ConsensusStateManager` i.e. `finalized_header` is said to be a header of a beacon block exclusively?

Reply:
1. this is exactly what i thought. did some research yesterday and i think it is necessary to also validate the body root as you stated. I am adding this right now.
2. there you go:
https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md#lightclientheader